README
$Id: README,v 1.8 1999/07/21 16:37:36 steve Exp $

Greetings,

    This file briefly describes the netCDFPerl package.

INTRODUCTION
------------

    The netCDFPerl package is a perl extension for accessing netCDF
    datasets based on version 2 of the netCDF package (netCDF-2).  For
    example, the following netCDF-2 actions:

        1.  Open dataset "foo.nc";

        2.  Write a 2 x 3 array of double-precision values into the
        starting corner of the first variable; and

        3.  Close the dataset.

    can be accomplished by the following C fragment:

        int     ncid = ncopen("foo.nc", NC_WRITE);
        int     varid = 0;
        long    start[2], count[2];
        double  values[6];

        ncid = ncopen("foo.nc", NC_WRITE);

        start[0] = 0;
        start[1] = 0;
        count[0] = 2;
        count[1] = 3;

        values[0] = 0.0;
        values[1] = 1.0;
        values[2] = 2.0;
        values[3] = 3.0;
        values[4] = 4.0;
        values[5] = 5.0;

        ncvarput(ncid, varid, start, count, values);

        ncclose(ncid);

    or by this equivalent perl fragment:

        use NetCDF;

        $ncid = NetCDF::open("foo.nc", NetCDF::WRITE);
        $varid = 0;
        @start = (0, 0);
        @count = (2, 3);
        @values = (0, 1, 2, 3, 4, 5);
        NetCDF::varput($ncid, $varid, \@start, \@count, \@values);
        NetCDF::close($ncid);

    There are perl-callable functions for all appropriate functions of
    the netCDF-2 API.

netCDF-2 vs. netCDF-3
---------------------

    Currently, the NetCDFPerl extension module is implemented using
    version 2 of the netCDF API (netCDF-2).  On 1997-05-16, version 3
    of the netCDF package was released (netCDF-3).  This newer version,
    however, contains a netCDF-2 backward compatibility interface.
    Thus, with a few minor changes already made, NetCDFPerl works with
    netCDF-3 as well as netCDF-2.

    Users should be aware, however, that the NetCDFPerl user-interface
    is based on netCDF-2.  In particular, the NetCDFPerl documentation
    doesn't describe the NetCDFPerl interface in detail but instead
    describes the differences between it and the netCDF-2 interface.
    The intention was that users would use the NetCDFPerl and netCDF-2
    documentation together in order to program using NetCDFPerl.

    With the deprecation of the netCDF-2 interface, this has become
    slightly problematical -- but not unworkable.  The solution is to
    ensure the availability of the netCDF-2 documentation.  Currently,
    the netCDF version 2 User's Guide is available at the following URL:

	ftp://ftp.unidata.ucar.edu/pub/netcdf/guide-2.4.3.ps.Z

    and the netCDF-library manual page for netCDF-2 is available as the
    file "netcdf-2.4.3/src/doc/netcdf.3" in the netCDF-2 distribution at
    the following URL:

	ftp://ftp.unidata.ucar.edu/pub/netcdf/netcdf-2.4.3.tar.Z


AVAILABILITY
------------

    netCDFPerl is freely available from either of the following URLs:

        http://www.unidata.ucar.edu/packages/netcdf-perl/
    
    or

        ftp://ftp.unidata.ucar.edu/pub/netcdf-perl/netcdf-perl.tar.Z

INSTALLATION
------------

    See the file INSTALL in the top-level directory of the netCDFPerl
    distribution for instructions on how to incorporate netCDFPerl into
    your perl utility.

    You will need write access to your installed perl(1) libraries in
    order to install netCDFPerl.

ADDITIONAL INFORMATION
----------------------

    See the installed manual page, netCDFPerl(1), for additional
    information.

    There is a WWW page for netCDFPerl.  It's URL is

        http://www.unidata.ucar.edu/packages/netcdf-perl

MAILING-LIST
------------

    There is a netCDFPerl mailing-list.  To subscribe, send to the
    following address:

        majordomo@unidata.ucar.edu

    a message whose body consists solely of the following:

        subscribe netcdf-perl [opt-addr]

    where [opt-addr] is an optional email address -- if you use it, then
    mailing-list postings will be sent to it; otherwise, they will be
    sent to the return address of the subscription request.

TRANSLATIONS
------------

    This page has been translated into Croatian. The URL is
    <http://science.webhostinggeeks.com/netcdf-perl">.


Regards,
Steve Emmerson <support@unidata.ucar.edu>


HISTORY
Release		Comments

1.2.4	2009-05-08
		Removed requirement for crediting UCAR/Unidata from 
		COPYRIGHT file.  Replaced with non-obligatory request.

		Eliminated some compiler warnings by adding a "default:"
		member to some "switch" statements.

		Verified correctness against netCDF-4 library.

		Eliminated MAX_NC_OPEN if the netCDF header-file doesn't
		have it.

1.2.3   2004-03-10

		Ported testing script "perl/test.pl" to perl v5.8.0.
		This should not have been necessary.

1.2.2	2002-01-09

		Ported to perl 5.005 and 5.6.0.

1.2.1	2000-07-25
		Improved "make install" with respect to installation of
		manual pages.

                Tried to clarify the fact that this extension-module is
                based on version 2 of the netCDF package.

		Added "POLLUTE=1" to the perl command that creates
		the makefile in the perl/ subdirectory to ensure
		compatibility with the new naming convensions of perl
		5.6.

1.2	1997-05-20
		Corrected manual page: fixed netcdf(3) reference so that
		man-cgi(1) will work with it.

                Ported to netCDF 3.  This required eliminating some
                netCDF C macros that should never have been in the
                public interface in the first place.  For example, I
                used NetCDF::RDWR in the test script "perl/test.pl"
                when I should have used NetCDF::NOWRITE.  Any perl
                script using these macros will have to be modified
                (fortunately, perl(1) reports undefined symbols).

                Improved INSTALL document: in particular, added
                commentary about modifying perl/Makefile.PL as
                necessary.

1.1	1996-08-22
		Allowed use of references for scalar output arguments.

                Documented lack of number-of-elements arguments in
                interface function and how to specify a scalar variable
                to NetCDF::vardef().

		Corrected configuration script so that setting of CC
		variable has an effect.

		Replaced use of croak() in perl/NetCDF.xs with warn().
		This will prevent the netCDF/Perl interface from
		aborting on error.

		Corrected typo in README file: changed `rcount' variable
		to `count'.

		Added check of NetCDF::attname() to perl `test' target.

		Slightly enhanced README.

		Added caveat about only using initialized output variables.

		Improved INSTALL file: added commentary about requiring
		position-independent netCDF object modules for a dynamic
		netCDFPerl module and listed desired information for
		diagnosing installation problems.

		Adapted perl/Makefile.PL.in to latest perl release.

1.0.4	1995-06-30
		Corrected names of netCDFPerl files in perl/MANIFEST file.

1.0.3	1995-06-29
		Fixed incorrect behavior of NetCDF::varput1().

		Converted top-level Makefile to POSIX.2 Standard.

1.0.2	1995-06-20
		Bug Fixes:
		    NetCDF::recget():
			First record was always returned regardless of
			requested record index.

		Documentation:
		    INSTALL File:
                        Reordered list of example, diagnostic commands
                        to be sent when a problem occurs and added `cat
                        VERSION`.

                        Added netCDF package to list of prerequisite
                        packages.

                        Added SunOS 5.4 to list systems on which
                        netCDFPerl has been successfully installed.

1.0.1	1995-06-12
		Added netCDFPerl URL to README.

		Fixed bug that caused a netCDF/perl record-variable size
		mismatch under SunOS 5 with the GNU C compiler.

		Fixed bug in NetCDF::getatt(): it used to crash the
		process.

1.0	1995-06-07
                Added CPP_NETCDF, LD_NETCDF, and LIBS examples to
		CUSTOMIZE file.

                Generalized netCDF::ncrecput and netCDF::ncrecget to
                allow references to scalar values.

		Added support for perl strings.  Reading a netCDF 
		character variable now returns a perl string rather than
		an array of integer values.

                Corrected bug in netCDF::ncattname: it would only
                return the first global attribute.  Thanks to Jon Meek
                <meekj@Cyanamid.COM>.

		Renamed everything in the API to conform to perl 5 
		conventions and to reduce redundancy:
		    *  The module name now starts with a capital;

                    *  Function names no longer contain an `nc' prefix;
		       and

                    *  Constant names no longer contain an `NC_'
		       substring.
		For example,
		    netCDF::nccreate($path, netCDF::NC_CLOBBER)
		is now
		    NetCDF::create($path, NetCDF::CLOBBER)

0.1	1995-05-23
		Ported installation mechanism to perl 5.001.

0.0	1995-05-15
		Initial release.


perl/test.pl
#!/usr/local/bin/perl

use warnings;
use strict 'vars';
use NetCDF;


#@a=(1,2);
#NetCDF::foo(\@a);


my $pathname = "foo.nc";
my $dim0name = "dim0";
my $dim1name = "dim1";
my $scalarvarname = "scalarvar";
my $fixvarname = "fixvar";
my $ndims = 4;
my $nvars = 5;
my $nrecvar = 3;
my $nregrecvar = 2;
my $natts = 0;
my @dimlen = (2, 3);
my $dimstrlen = 12;
my $dimstrname = "strlen";
my $scalarvarvalue = 5.5;

my $attid;
my $attval;
my $dim0id;
my $dim0len;
my $dim1id;
my $dim1len;
my $dimid;
my $dimstrid;
my $scalarvarid;
my $scalarvartype;
my $fixvarid;
my $fixvarnatts;
my $fixvarndims;
my $fixvartype;
my $globattname;
my $histatt;
my $i;
my $idim;
my $j;
my $len;
my $length;
my $na;
my $name;
my $ncid;
my $nd;
my $nrecvars;
my $nv;
my $recdimid;
my $recvar0id;
my $recvar1id;
my $recvars;
my $recvarstrid;
my $status;
my $type;
my $value;
my $varid;
my $varref;
my @atts;
my @coords;
my @dimids;
my @counts;
my @scalarvarcount;
my @fixvarcount;
my @fixvardimids;
my @fixvarstart;
my @fixvarvalue;
my @record;
my @recsizes;
my @recvar0;
my @recvar1;
my @recvarids;
my @recvars;
my @values;


#
# Create netCDF file.
#
print STDERR "Creating netCDF file...........................";
$ncid = NetCDF::create($pathname, NetCDF::CLOBBER);
die "Couldn't open netCDF file\n" if $ncid < 0;
print STDERR "ok\n";


#
# Set fill mode.
#
print STDERR "Setting fill mode..............................";
NetCDF::setfill($ncid, NetCDF::NOFILL) == 0 ||
    die "Couldn't set fill mode\n";
print STDERR "ok\n";


#
# Define scalar variable.
#
print STDERR "Defining scalar variable.......................";
$scalarvartype = NetCDF::FLOAT;
@dimids = ();
$scalarvarid = NetCDF::vardef($ncid, $scalarvarname, $scalarvartype, \@dimids);
die "Couldn't define scalar variable\n" if $scalarvarid < 0;
print STDERR "ok\n";


#
# Define fixed dimensions.
#
print STDERR "Defining fixed dimensions......................";
$dim0len = $dimlen[0];
$dim0id = NetCDF::dimdef($ncid, $dim0name, $dim0len);
die "Couldn't define first dimension\n" if $dim0id < 0;
print STDERR "ok\n";

@dimids = ( $dim0id );

$dim1len = $dimlen[1];
$dim1id = NetCDF::dimdef($ncid, $dim1name, $dim1len);
die "Couldn't define second dimension\n" if $dim1id < 0;

push(@dimids, $dim1id);

$dimstrid = NetCDF::dimdef($ncid, $dimstrname, $dimstrlen);
die "Couldn't define string dimension\n" if $dimstrid < 0;


#
# Define fixed variables.
#
print STDERR "Defining fixed variables.......................";
$fixvartype = NetCDF::FLOAT;
$fixvarndims = @dimids;
$fixvarid = NetCDF::vardef($ncid, $fixvarname, $fixvartype, \@dimids);
die "Couldn't define fixed variable\n" if $fixvarid < 0;
@fixvardimids = @dimids;
print STDERR "ok\n";


#
# Put global attribute.
#
print STDERR "Writing global attribute.......................";
$globattname = "history";
$histatt = "Created by $0 on " . localtime();
$attid = NetCDF::attput($ncid, NetCDF::GLOBAL, $globattname, NetCDF::CHAR,
			$histatt);
$attid != -1 || die "Couldn't write global attribute\n";
print STDERR "ok\n";


#
# Put variable attributes.
#
print STDERR "Writing variable attributes....................";
@atts = (255, -128);
$attid = NetCDF::attput($ncid, $fixvarid, "att_byte", NetCDF::BYTE,
			  \@atts);
die "Couldn't put byte attribute\n" if $attid < 0;

$attid = NetCDF::attput($ncid, $fixvarid, "att_char", NetCDF::CHAR,
			  "string");
die "Couldn't put char attribute\n" if $attid < 0;

@atts = (5, 6, 7);
$attid = NetCDF::attput($ncid, $fixvarid, "att_short", NetCDF::SHORT,
			  \@atts);
die "Couldn't put short attribute\n" if $attid < 0;

@atts = (3, 4);
$attid = NetCDF::attput($ncid, $fixvarid, "att_long", NetCDF::LONG,
			  \@atts);
die "Couldn't put long attribute\n" if $attid < 0;

$attid = NetCDF::attput($ncid, $fixvarid, "att_float", NetCDF::FLOAT, 
			  2.7182818);
die "Couldn't put float attribute\n" if $attid < 0;

@atts = (2.7182818, 3.1415927);
$attid = NetCDF::attput($ncid, $fixvarid, "att_double", NetCDF::DOUBLE, 
			  \@atts);
die "Couldn't put double attribute\n" if $attid < 0;
print STDERR "ok\n";

$fixvarnatts = 6;


#
# Define record dimension.
#
print STDERR "Defining record dimension......................";
$recdimid = NetCDF::dimdef($ncid, "recdim", UNLIMITED);
die "Couldn't define record dimension\n" if $recdimid < 0;
print STDERR "ok\n";


#
# Define record variables.
#
print STDERR "Defining record variables......................";
$recvar0id = NetCDF::vardef($ncid, "recvar0", NetCDF::SHORT, 
			     [$recdimid, $dim0id]);
die "Couldn't define first record variable\n" if $recvar0id < 0;

$recvar1id = NetCDF::vardef($ncid, "recvar1", NetCDF::FLOAT, 
			     [$recdimid, $dim1id]);
die "Couldn't define second record variable\n" if $recvar1id < 0;

$recvarstrid = NetCDF::vardef($ncid, "recvarstr", NetCDF::CHAR, 
			     [$recdimid, $dimstrid]);
die "Couldn't define string record variable\n" if $recvarstrid < 0;

print STDERR "ok\n";


#
# End definition.
#
print STDERR "Ending definition..............................";
$status = NetCDF::endef($ncid);
die "Couldn't end definition\n" if $status < 0;
print STDERR "ok\n";


#
# Write value of scalar variable.
#
print STDERR "Writing scalar variable value..................";
@coords = ();
@counts = (1);
$status = NetCDF::varput($ncid, $scalarvarid, \@coords, \@counts,
			   $scalarvarvalue);
die "Couldn't write scalar-variable\n" if $status < 0;
print STDERR "ok\n";


#
# Write values to fixed variable.
#
print STDERR "Writing fixed variable values..................";
@fixvarstart = (0, 1);
@fixvarcount = (2, 1);
@fixvarvalue = (998, 999);
$status = NetCDF::varput($ncid, $fixvarid, \@fixvarstart, \@fixvarcount,
			   \@fixvarvalue);
die "Couldn't write fixed-variable\n" if $status < 0;
print STDERR "ok\n";

##@arg = ();
#$argref = [];
#$status = foo($argref);
#print STDERR "foo() = $status\n";
#print STDERR "argref = @$argref\n";
##print STDERR "arg = @arg\n";


#
# Synchronize netCDF file.
#
print STDERR "Synchronizing netCDF I/O.......................";
$status = NetCDF::sync($ncid);
die "Couldn't synchronize netCDF file\n" if $status < 0;
print STDERR "ok\n";


#
# Write values to record variables.
#
print STDERR "Writing record values..........................";
@recvar0 = (101 .. (100+$dim0len));
@recvar1 = (201 .. (200+$dim1len));
$status = NetCDF::recput($ncid, 0, [\@recvar0, \@recvar1, \"hello world\0"]);
die "Couldn't write to record variables\n" if $status < 0;
print STDERR "ok\n";


#
# Save the values of the record variables.
#
$recvars = [ \@recvar0, \@recvar1 ];


#
# Close netCDF file.
#
print STDERR "Closing netCDF file............................";
$status = NetCDF::close($ncid);
die "Couldn't close netCDF file\n" if $status < 0;
print STDERR "ok\n";


#
# Open netCDF file.
#
print STDERR "Opening netCDF file for reading................";
$ncid = NetCDF::open($pathname, NOWRITE);
die "Couldn't open netCDF file\n" if $status < 0;
print STDERR "ok\n";


#
# Inquire about netCDF file.
#
print STDERR "Inquiring about netCDF file....................";
$status = NetCDF::inquire($ncid, $nd, $nv, $na, $dimid);
die "Couldn't inquire about netCDF file\n" if $status < 0;
die "Incorrect netCDF information\n" if $nd != $ndims;
die "Incorrect netCDF information\n" if $nv != $nvars;
die "Incorrect netCDF information\n" if $dimid != $recdimid;

#die "Incorrect netCDF information\n" if $nd != $ndims ||
#					$nv != $nvars ||
#					$na != 1 ||
#					$dimid != $recdimid;
print STDERR "ok\n";


#
# Get global attribute.
#
print STDERR "Checking global attribute name.................";
NetCDF::attname($ncid, NetCDF::GLOBAL, 0, $name) == 0 ||
    die "Couldn't get global attribute name\n";
($name eq $globattname) ||
    die "Incorrect global attribute name: \"$name\" != \"$globattname\"\n";
print STDERR "ok\n";
print STDERR "Checking global attribute value................";
$attval = "";
NetCDF::attget($ncid, NetCDF::GLOBAL, "history", \$attval) == 0 ||
    die "Couldn't get fixed variable attribute\n";
$attval =~ /$histatt/ || 
    die "Incorrect global attribute value: \"$attval\"\n";
print STDERR "ok\n";

#
# Vet scalar variable.
#
print STDERR "Checking scalar variable.......................";
NetCDF::varinq($ncid, $scalarvarid, $name, $type, $ndims, \@dimids, $natts)
    == 0 || die "Couldn't get information on scalar variable\n";
if ($name ne $scalarvarname ||
    $type != $scalarvartype ||
    $ndims != 0 ||
    $natts != 0)
{
    die "Incorrect fixed variable information\n";
}
@coords = ();
@counts = ();
@values = ();
NetCDF::varget($ncid, $scalarvarid, \@coords, \@counts, \@values) == 0 ||
    die "Couldn't get value of scalar variable\n";
$values[0] == $scalarvarvalue ||
    die "Incorrect value for scalar variable\n";
print STDERR "ok\n";



#
# Get ID of second dimension.
#
print STDERR "Getting second dimension ID....................";
$dimid = NetCDF::dimid($ncid, $dim1name);
die "Incorrect second dimension ID\n" if $dimid != $dim1id;
print STDERR "ok\n";


#
# Vet second dimension information.
#
print STDERR "Checking second dimension information..........";
NetCDF::diminq($ncid, $dim1id, $name, $length) == 0 ||
    die "Couldn't get information on second dimension\n";
($name eq $dim1name) || 
    die "Incorrect second dimension name: \"$name\" != \"$dim1name\"\n";
($length == $dim1len) ||
    die "Incorrect second dimension length: $length != $dim1len\n";
print STDERR "ok\n";


#
# Vet variable ID of fixed variable.
#
print STDERR "Checking fixed variable ID.....................";
$varid = NetCDF::varid($ncid, $fixvarname);
($varid >= 0) || die "Couldn't get variable ID of fixed variable\n";
($varid == $fixvarid) || die "Incorrect fixed variable ID: ",
			     "$varid != $fixvarid\n";
print STDERR "ok\n";


#
# Vet information on fixed variable.
#
printf STDERR "Checking fixed variable information............";
@dimids = ();
NetCDF::varinq($ncid, $fixvarid, $name, $type, $ndims, \@dimids, $natts)
    == 0 || die "Couldn't get information on fixed variable\n";
if ($name ne $fixvarname ||
    $type != $fixvartype ||
    $ndims != $fixvarndims ||
    $natts != $fixvarnatts)
{
    die "Incorrect fixed variable information\n";
}
for ($idim = 0; $idim < $ndims; $idim++)
{
    ($dimids[$idim] == $fixvardimids[$idim]) ||
	die "Incorrect dimension $idim ID: ",
	    "$dimids[$idim] != $fixvardimids[$idim]\n";
}
print STDERR "ok\n";


#
# Vet last value of first record variable.
#
#
# For some reason, the following assignment results in a segmentation
# violation in the "varget1" statement unless the "print" statement is enabled
# when being interpreted by perl v5.8.0 (earlier version worked OK).
#
# @coords = (0, $#recvar0);
# print STDERR "ncid=$ncid; recvar0id=$recvar0id; coords=(@coords)\n";
#
# Hence, we use this equivalent assignment, instead.
#
@coords = (0, scalar(@recvar0)-1);
print STDERR "Checking last value of first record variable...";
NetCDF::varget1($ncid, $recvar0id, \@coords, $value) == 0 ||
    die "Couldn't get last value of first record variable\n";
$value == $recvar0[$#recvar0] ||
    die "Incorrect last value of first record variable: ",
	"$value != $recvar0[$#recvar0]\n";
print STDERR "ok\n";


#
# Vet fixed variable values.
#
print STDERR "Reading fixed variable values..................";
@values = ();
$status = NetCDF::varget($ncid, $varid, \@fixvarstart, \@fixvarcount,
			   \@values);
($status == 0) || die "Couldn't read from fixed variable\n";
if (@fixvarvalue != @values)
{
    die "Incorrect vector size\n";
}
else
{
    for ($i = 0; $i < @fixvarvalue; $i++)
    {
	$fixvarvalue[$i] == $values[$i] ||
	    die "Incorrect fixed value $i: ",
		"$fixvarvalue[$i] != $values[$i]\n";
    }
}
print STDERR "ok\n";


#
# Vet fixed variable attributes.
#
print STDERR "Reading fixed variable attributes..............";
NetCDF::attname($ncid, $fixvarid, 0, $name) == 0 ||
    die "Couldn't get attribute name\n";
$name eq "att_byte" || die "Incorrect attribute name: \"$name\"\n";
NetCDF::attinq($ncid, $fixvarid, "att_float", $type, $len) == 0 ||
    die "Couldn't get information on fixed variable attribute\n";
$type == NetCDF::FLOAT || die "Incorrect attribute type\n";
$len == 1 || die "Incorrect attribute length\n";
@values = ();
NetCDF::attget($ncid, $fixvarid, "att_float", \@values) == 0 ||
    die "Couldn't get fixed variable attribute\n";
abs(($values[0] - 2.7182818) / 2.7182818) < .000001 || 
    die "Incorrect attribute value: $values[0]\n";
print STDERR "ok\n";


#
# Vet NetCDF::typelen().
#
print STDERR "Checking typelen().............................";
NetCDF::typelen(NetCDF::FLOAT) == 4 || die "Incorrect float length\n";
print STDERR "ok\n";


#
# Inquire about record variables.
#
@recvarids = ();
@recsizes = ();
NetCDF::recinq($ncid, $nrecvars, \@recvarids, \@recsizes) == 0 ||
    die "Couldn't inquire about record variables\n";
print STDERR "Checking number of record variables............";
$nrecvars == $nrecvar || die "Incorrect number of record variables: ",
			     "$nrecvars != $nrecvar\n";
print STDERR "ok\n";
print STDERR "Checking variable IDs..........................";
($recvarids[0] == $recvar0id && $recvarids[1] == $recvar1id) ||
    die "Incorrect record variable IDs: ",
	"($recvarids[0],$recvarids[1]) != ($recvar0id,$recvar1id)\n";
print STDERR "ok\n";
print STDERR "Checking variable sizes........................";
$recsizes[0] == $dim0len * NetCDF::typelen(NetCDF::SHORT) &&
    $recsizes[1] == $dim1len * NetCDF::typelen(NetCDF::FLOAT) ||
    die "Incorrect record variable sizes\n";
print STDERR "ok\n";

#
# Read values of record variables.
#
print STDERR "Reading values of record variables.............";
@record = ();
$status = NetCDF::recget($ncid, 0, \@record);
($status >= 0) || die "Couldn't read record\n";
$nv = @record;
$nv == $nrecvar ||
    die "Incorrect number of record variables: $nv != $nrecvar\n";
for ($i = 0; $i < $nregrecvar; $i++)
{
    $varref = $record[$i];
    (@$varref == $dimlen[$i]) || 
	die "Incorrect number of elements in record variable $i: ",
	    "@$varref != $dimlen[$i]\n";
    for ($j = 0; $j < $dimlen[$i]; $j++)
    {
	($$varref[$j] == $$recvars[$i][$j]) ||
	    die "Incorrect record variable value: ",
		"$$varref[$j] != $recvars[$i][$j]\n";
    }
}
${$record[2]} =~ /hello world/ || 
    die "Incorrect record string-variable\n";
print STDERR "ok\n";


print STDERR "Closing netCDF file............................";
NetCDF::close($ncid) == 0 || die "Couldn't close netCDF file\n";
print STDERR "ok\n";


perl/AR.pl
#!/usr/bin/perl

use warnings;
use strict 'vars';
use NetCDF;

my $l_file;
my $l_varId;
my $l_name;
my $l_dimCount;
my @l_dimIds;
my $l_atts;
my $l_size;
my @l_fileValues;
my $l_datatype;

$l_file = NetCDF::open("AR_LATEST_PR_CT_MYO_AR_58AA_20110519.nc", NOWRITE);
$l_file >= 0 || die "open() failue";
                
$l_varId = NetCDF::varid($l_file, "LONGITUDE");
$l_varId >= 0 || die "varid() failure";

NetCDF::varinq($l_file, $l_varId, \$l_name, \$l_datatype, \$l_dimCount,
        \@l_dimIds, \$l_atts) == 0 || die "varinq() failure";

NetCDF::diminq($l_file, $l_dimIds[0], \$l_name, \$l_size) == 0 ||
        die "diminq() failure";
NetCDF::varget($l_file, $l_varId, [0], [$l_size], \@l_fileValues) == 0 ||
        die "varget() failure";

foreach my $l_value (@l_fileValues) {
    print "$l_value, Value $l_fileValues[$l_value]\n";
}


perl/setfill.pl
#!/usr/local/bin/perl
##!/data/zender/bin/SUNMP/perl -w

use NetCDF;

# Set output flushing to help debugging on hard crashes. 
# These options update the filehandle after every output statement.
# See the camel book, p. 110.
select((select(STDOUT),$|=1)[0]);
select((select(STDERR),$|=1)[0]);

$out_file='foo.nc';

# Dimensions
%dim_cnt=(
	  'wv_nbr',NetCDF::UNLIMITED,
	  );

# Variables
%var_dim=(
		'HWHM','wv_nbr',
		'HWHM_t_dep','wv_nbr',
		'iso_nbr','scalar',
		'ln_str','wv_nbr',
		'mlc_nbr','scalar',
		'nrg_lo','wv_nbr',
		'wv_nbr','wv_nbr',
		);
%var_type=(
		'HWHM',NetCDF::FLOAT,
		'HWHM_t_dep',NetCDF::FLOAT,
		'iso_nbr',NetCDF::SHORT,
		'ln_str',NetCDF::FLOAT,
		'mlc_nbr',NetCDF::SHORT,
		'nrg_lo',NetCDF::FLOAT,
		'wv_nbr',NetCDF::FLOAT,
		);

# Create netCDF output file
print STDOUT 'Creating netCDF file...';
$nc_id=NetCDF::create($out_file,NetCDF::CLOBBER);
die "Could not open netCDF file\n" if $nc_id < 0;
print STDOUT "ok\n";

# Set fill mode
print STDOUT 'Setting fill mode...';
NetCDF::setfill($nc_id,NetCDF::NOFILL) == 0 || die "Could not set fill mode\n";
print STDOUT "ok\n";

# Define dimensions
print STDOUT 'Defining dimensions...';
foreach $dim_nm (keys(%dim_cnt)){
    $dim_id{$dim_nm}=NetCDF::dimdef($nc_id,$dim_nm,$dim_cnt{$dim_nm});
    die "Could not define dimension\n" if $dim_id{$dim_nm} < 0;
} # end loop over dimensions
print STDOUT "ok\n";

# Now that we have the dimension IDs, form the dimension ID vectors
@scalar=();
%dim_vec=(
	  'scalar',\@scalar,
	  'wv_nbr',$dim_id{'wv_nbr'},
	  );

# Define variables
print STDOUT 'Defining variables...';
foreach $var_nm (sort keys(%var_dim)){

    $var_id{$var_nm}=NetCDF::vardef($nc_id,$var_nm,$var_type{$var_nm},\$dim_vec{$var_dim{$var_nm}});
    die "Could not define variable\n" if $var_id{$var_nm} < 0;

} # end loop over variables
print STDOUT "ok\n";

# End definition
print STDOUT 'Ending definition mode...';
$status=NetCDF::endef($nc_id);
die "Could not end definition mode\n" if $status < 0;
print STDOUT "ok\n";

# Close netCDF file
print STDOUT 'Closing netCDF file...';
$status=NetCDF::close($nc_id);
die "Could not close netCDF file\n" if $status < 0;
print STDOUT "ok\n";


perl/tst_chunks.pl
use NetCDF;

#
# Open netCDF file.
#
print "Opening netCDF file for reading\n";
$ncid = NetCDF::open("tst_chunks.nc", NetCDF::NOWRITE);
die "Couldn't open netCDF file\n" if $ncid < 0;

#
# Inquire about netCDF file.
#
print "Inquiring about netCDF file\n";
$status = NetCDF::inquire($ncid, $nd, $nv, $na, $dimid);
die "Couldn't inquire about netCDF file\n" if $status < 0;

print "nd: $nd  nv: $nv  na: $na  dimid: $dimid\n";

#
# Check the value of a variable element.
#
print "Checking variable value: ";
@start = (1, 2, 3);
@count = (1, 1, 1);
@values = ();
$status = NetCDF::varget($ncid, 0, \@start, \@count, \@values);
($status == 0) || die "Couldn't read from fixed variable\n";
$values[0] == 92 || die "value not 92: $values[0]\n"; 
print "correct\n";

NetCDF::close($ncid) == 0 || die "Couldn't close netCDF file\n";


perl/rec.pl
#!/usr/local/bin/perl

use NetCDF;


#
# Create netCDF file.
#
print STDERR "Creating netCDF file...........................";
$ncid = NetCDF::create("rec.nc", NetCDF::CLOBBER);
die "Couldn't open netCDF file\n" if $ncid < 0;
print STDERR "ok\n";


#
# Set fill mode.
#
print STDERR "Setting fill mode..............................";
NetCDF::setfill($ncid, NetCDF::NOFILL) == 0 ||
    die "Couldn't set fill mode\n";
print STDERR "ok\n";


#
# Define record dimension.
#
print STDERR "Defining record dimension......................";
$recdimid = NetCDF::dimdef($ncid, "recdim", UNLIMITED);
die "Couldn't define record dimension\n" if $recdimid < 0;
print STDERR "ok\n";


#
# Define record variables.
#
print STDERR "Defining record variables......................";
$recvar0id = NetCDF::vardef($ncid, "recvar0", NetCDF::SHORT, 
			     [$recdimid]);
die "Couldn't define first record variable\n" if $recvar0id < 0;

$recvar1id = NetCDF::vardef($ncid, "recvar1", NetCDF::FLOAT, 
			     [$recdimid]);
die "Couldn't define second record variable\n" if $recvar1id < 0;
print STDERR "ok\n";


#
# End definition.
#
print STDERR "Ending definition..............................";
$status = NetCDF::endef($ncid);
die "Couldn't end definition\n" if $status < 0;
print STDERR "ok\n";


#
# Synchronize netCDF file.
#
print "Synchronizing netCDF I/O.......................";
$status = NetCDF::sync($ncid);
die "Couldn't synchronize netCDF file\n" if $status < 0;
print "ok\n";


#
# Write values to record variables.
#
print STDERR "Writing record values..........................";
$recvar0 = 10;
$recvar1 = 20;
$status = NetCDF::recput($ncid, 0, [\$recvar0, \$recvar1]);
die "Couldn't write to record variables\n" if $status < 0;
$recvar0 = 11;
$recvar1 = 21;
$status = NetCDF::recput($ncid, 1, [\$recvar0, \$recvar1]);
die "Couldn't write to record variables\n" if $status < 0;
print STDERR "ok\n";


#
# Close netCDF file.
#
print STDERR "Closing netCDF file............................";
$status = NetCDF::close($ncid);
die "Couldn't close netCDF file\n" if $status < 0;
print STDERR "ok\n";


#
# Reopen netCDF file.
#
print STDERR "Reopening netCDF file..........................";
$ncid = NetCDF::open("rec.nc", NetCDF::NOWRITE);
die "Couldn't open netCDF file\n" if $ncid < 0;
print STDERR "ok\n";


#
# Read values from record variables.
#
#print STDERR "Reading record values..........................";
#$ref = [\@var0, \@var1];
#$status = NetCDF::recget($ncid, 0, $ref);
$status = NetCDF::recget($ncid, 1, [\@var0, \@var1]);
#$var0ref = \@var0;
#$var1ref = \@var1;
#$status = NetCDF::recget($ncid, 0, [$var0ref, $var1ref]);
#$ref = [$var0ref, $var1ref];
#$status = NetCDF::recget($ncid, 0, $ref);
#die "Couldn't read from record variables\n" if $status < 0;
#if ($var0[0] != 11 || $var1[0] != 21)
#{
#    print STDERR "variable 0 = $var0[0]\n";
#    print STDERR "variable 1 = $var1[0]\n";
#    die "Incorrect record values\n";
#}
#print STDERR "ok\n";


##
## Read values from record variables.
##
#print STDERR "Reading record values..........................";
#@record = ();
#$status = NetCDF::recget($ncid, 0, \@record);
#die "Couldn't read from record variables\n" if $status < 0;
#$nv = @record;
#$nv == 2 || die "Incorrect number of record variables: $nv != 2\n";
#$var0ref = $record[0];
#$var1ref = $record[1];
#if ($$var0ref[0] != 0 || $$var1ref[0] != 1)
#{
#    print STDERR "variable 0 = $$var0ref[0]\n";
#    print STDERR "variable 1 = $$var1ref[0]\n";
#    die "Incorrect record values\n";
#}
#print STDERR "ok\n";


#
# Read values from record variables.
#
#print STDERR "Reading record values..........................";
#$status = NetCDF::recget($ncid, 0, [\$var0, \$var1]);
#die "Couldn't read from record variables\n" if $status < 0;
#if ($var0 != 0 || $var1 != 1)
#{
#    print STDERR "variable 0 = $var0\n";
#    print STDERR "variable 1 = $var1\n";
#    die "Incorrect record values\n";
#}
#print STDERR "ok\n";


#
# Write new values to record variables.
#
print STDERR "Writing record values..........................";
$recvar0 = 100;
$recvar1 = 200;
$status = NetCDF::recput($ncid, 0, [\$recvar0, \$recvar1]);
die "Couldn't write to record variables\n" if $status < 0;
$recvar0 = 110;
$recvar1 = 210;
$status = NetCDF::recput($ncid, 1, [\$recvar0, \$recvar1]);
die "Couldn't write to record variables\n" if $status < 0;
print STDERR "ok\n";


#
# Close netCDF file.
#
print STDERR "Closing netCDF file............................";
$status = NetCDF::close($ncid);
die "Couldn't close netCDF file\n" if $status < 0;
print STDERR "ok\n";
